### DISCUSSION TITLE
AsyncValue.guard error 'side effect'

### DISCUSSION THREAD

**Discussion by josh-burton** [rrousselGit/riverpod#794] [Q&A]
I really like AsyncValue.guard for it's simplicity in constructing an AsyncValue from a future, but I find myself needing a callback if there was an error so I can log the error to crash reporting etc.

Currently after the guard I'm doing this:

```
myAsyncValue.whenOrNull(error: (e, stack, previous) {
      Fimber.e("Failed to load data", ex: e);
    });
```

I think this works ok but it doesn't quite feel right. Maybe a `myAsyncValue.whenError()` function would feel more natural.

**Comment by rrousselGit** [comment]
You can alternatively do:

```dart
if (myAsyncValue is AsyncError)
   Fimber.e("Failed to load data", ex: myAsyncValue.error);

```

**Reply by josh-burton** [reply]
Would you be open to a pull request to add a `whenError` method?

**Reply by rrousselGit** [reply]
I'm not sure that this is something I'd like to recommend folks to do. For now I'd suggest using an extension
